File: Semestre_2/Vision_Computadora/Entregas/Matlab/T2_15_Transformaciones_Color.m
clear;
close all;
clc;

[file,path] = uigetfile({'*.jpg';'*.png';'*.jpeg';'*.bmp';'*.gif'},'File Selector - Imagen A'); 
ruta=strcat(path,file);
A=imread(ruta);
A= im2double(A);

% Matrices para almacenar nuevas imagenes
[ax ay az] = size(A);
gris_prom = zeros(ax, ay);
gris_pres = zeros(ax, ay);
gris_matlab = rgb2gray(A);

% Canales RGB
A_R = A(:,:,1);
A_G = A(:,:,2);
A_B = A(:,:,3);

% Transformación a escala de grises media y precisa
for m=1:ax
    for n=1:ay
        gris_prom(m,n)=(A_R(m,n) + A_G(m,n) + A_B(m,n))/3.0;
        gris_pres(m,n)=0.21*A_R(m,n) + 0.72*A_G(m,n) + 0.07*A_B(m,n);
    end
end

figure,
subplot(221);imshow(A);title('Imagen original');
subplot(222);imshow(gris_prom);title('Gris promedio');
subplot(223);imshow(gris_pres);title('Gris precisa');
subplot(224);imshow(gris_matlab);title('Gris Matlab');

File: Semestre_2/Vision_Computadora/Entregas/Python/T2_15_Transformaciones_Color.py
import matplotlib.pyplot as plt
import numpy as np
from skimage.io import imread
from tkinter import Tk, filedialog
from skimage.transform import resize

file_path1 = filedialog.askopenfilename(
    title='File Selector - Imagen A',
    filetypes=[("Images", "*.jpg;*.png;*.jpeg;*.bmp;*.gif")]
)
A = imread(file_path1).astype(float) / 255.0

file_path2 = filedialog.askopenfilename(
    title='File Selector - Imagen B',
    filetypes=[("Images", "*.jpg;*.png;*.jpeg;*.bmp;*.gif")]
)
B = imread(file_path2).astype(float) / 255.0

if A.ndim == 3 and A.shape[-1] == 4:
    A = A[..., :3]
if B.ndim == 3 and B.shape[-1] == 4:
    B = B[..., :3]

bx, by, bz = B.shape
A = resize(A, (bx, by, bz), preserve_range=True)

Rmax = np.maximum(A, B)
Rmin = np.minimum(A, B)

plt.figure()
plt.subplot(2,2,1); plt.imshow(A); plt.title('A'); plt.axis('off')
plt.subplot(2,2,3); plt.imshow(B); plt.title('B'); plt.axis('off')
plt.subplot(2,2,2); plt.imshow(Rmin); plt.title('Min(A,B)'); plt.axis('off')
plt.subplot(2,2,4); plt.imshow(Rmax); plt.title('Max(A,B)'); plt.axis('off')
plt.show()
